File: linux/fieldcommand/replay.py
"""Replays: a recorded game is its seed, its setup and every command it received, by tick. Fed back into a
fresh simulation that gives exactly the same game, which is what makes replays small — a few kilobytes for
an hour — and what makes them a bug report's best friend."""
import json
import os
import time

from .save import saves_dir

REPLAY_FORMAT = 1


class Replay:
    def __init__(self, data):
        self.data = data
        self.map = data["map"]
        self.players = data["players"]
        self.difficulty = int(data["difficulty"])
        self.seed = int(data["seed"])
        self.mission = data.get("mission")
        self.viewer = int(data.get("viewer", 0))
        self.commands = [(int(t), int(slot), cmd) for t, slot, cmd in data["commands"]]
        self.next = 0
        self.elapsed = float(data.get("elapsed", 0))
        self.label = data.get("label", "")


def replays_dir():
    return os.path.join(os.path.dirname(saves_dir()), "replays")


def path_for(name):
    return os.path.join(replays_dir(), f"{name}.json")


def replay_to_dict(world, viewer=0, label=""):
    from . import __version__
    return {"game": "field-command", "format": REPLAY_FORMAT, "version": __version__, "label": label,
            "saved_at": int(time.time()), "map": world.map["id"], "difficulty": world.difficulty.index,
            "seed": world.seed, "mission": world.mission.id if world.mission else None, "viewer": viewer,
            "players": [{"slot": p.slot, "name": p.name, "team": p.team, "ai": p.is_ai} for p in world.players.values()],
            "elapsed": world.elapsed, "ticks": world.tick,
            "winner_team": world.winner_team,
            "commands": [[t, s, c] for t, s, c in world.record]}


def write_replay(world, name="last", viewer=0, label=""):
    os.makedirs(replays_dir(), exist_ok=True)
    path = path_for(name)
    tmp = path + ".tmp"
    with open(tmp, "w") as f:
        json.dump(replay_to_dict(world, viewer, label), f, separators=(",", ":"))
    os.replace(tmp, path)
    return path


def read_replay(name="last"):
    with open(path_for(name)) as f:
        data = json.load(f)
    if data.get("game") != "field-command" or data.get("format") != REPLAY_FORMAT:
        raise ValueError("not a Field Command replay this version can read")
    return Replay(data)


def list_replays():
    """(name, label, saved_at, elapsed), newest first."""
    out = []
    try:
        names = os.listdir(replays_dir())
    except OSError:
        return out
    for fn in names:
        if not fn.endswith(".json"):
            continue
        try:
            with open(os.path.join(replays_dir(), fn)) as f:
                d = json.load(f)
            out.append((fn[:-5], d.get("label", ""), int(d.get("saved_at", 0)), float(d.get("elapsed", 0))))
        except (OSError, ValueError):
            continue
    return sorted(out, key=lambda r: -r[2])

File: macos/Sources/FieldCommand/Replay.swift
import Foundation

/// A recorded game: its seed, its setup and every command people sent, by tick — the same JSON document the
/// Linux edition's replay.py writes. Fed back into a fresh simulation it gives exactly the same game.
final class Replay {
    static let format = 1
    let map: String
    let players: [(slot: Int, name: String, team: Int, ai: Bool)]
    let difficulty: Int
    let seed: UInt64
    let mission: String?
    let viewer: Int
    let commands: [(Int, Int, [Any])]
    let elapsed: Double
    let label: String
    var next = 0

    struct Unreadable: Error, CustomStringConvertible { let description: String }

    init(_ d: [String: Any]) throws {
        guard jStr(d["game"]) == "field-command", jInt(d["format"]) == Replay.format else {
            throw Unreadable(description: "not a Field Command replay this version can read")
        }
        map = jStr(d["map"])
        players = jArr(d["players"]).map { jDict($0) }.map { (jInt($0["slot"]), jStr($0["name"]), jInt($0["team"]), SaveGame.jBoolean($0["ai"])) }
        difficulty = jInt(d["difficulty"])
        seed = UInt64(max(1, jInt(d["seed"])))
        mission = d["mission"] as? String
        viewer = jInt(d["viewer"])
        commands = jArr(d["commands"]).map { jArr($0) }.compactMap { $0.count >= 3 ? (jInt($0[0]), jInt($0[1]), jArr($0[2])) : nil }
        elapsed = Double(jNum(d["elapsed"]))
        label = jStr(d["label"])
    }

    static var replaysDir: URL { SaveGame.savesDir.deletingLastPathComponent().appendingPathComponent("replays") }
    static func path(_ name: String) -> URL { replaysDir.appendingPathComponent("\(name).json") }

    static func encode(_ w: SWorld, viewer: Int, label: String = "") -> [String: Any] {
        ["game": "field-command", "format": format, "version": appVersion, "label": label,
         "saved_at": Int(Date().timeIntervalSince1970), "map": jStr(w.map["id"]), "difficulty": w.difficulty.rawValue,
         "seed": Int(w.seed), "mission": w.mission.map { $0.id as Any } ?? NSNull(), "viewer": viewer,
         "players": w.players.values.sorted { $0.slot < $1.slot }.map {
             ["slot": $0.slot, "name": $0.name, "team": $0.team, "ai": $0.isAI] as [String: Any] },
         "elapsed": w.elapsed, "ticks": w.tick, "winner_team": w.winnerTeam.map { $0 as Any } ?? NSNull(),
         "commands": w.record.map { [$0.0, $0.1, $0.2] as [Any] }]
    }

    @discardableResult
    static func write(_ w: SWorld, name: String = "last", viewer: Int = 0, label: String = "") throws -> URL {
        try FileManager.default.createDirectory(at: replaysDir, withIntermediateDirectories: true)
        let data = try JSONSerialization.data(withJSONObject: encode(w, viewer: viewer, label: label))
        let url = path(name)
        try data.write(to: url, options: .atomic)
        return url
    }

    static func read(_ name: String = "last") throws -> Replay {
        let data = try Data(contentsOf: path(name))
        guard let d = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw Unreadable(description: "not JSON") }
        return try Replay(d)
    }

    /// (name, label, savedAt, elapsed), newest first.
    static func list() -> [(String, String, Int, Double)] {
        guard let names = try? FileManager.default.contentsOfDirectory(atPath: replaysDir.path) else { return [] }
        var out: [(String, String, Int, Double)] = []
        for fn in names where fn.hasSuffix(".json") {
            guard let data = try? Data(contentsOf: replaysDir.appendingPathComponent(fn)),
                  let d = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }
            out.append((String(fn.dropLast(5)), jStr(d["label"]), jInt(d["saved_at"]), Double(jNum(d["elapsed"]))))
        }
        return out.sorted { $0.2 > $1.2 }
    }
}
